### PR TITLE
remake keybinding ui and add mouse support

### LINKED ISSUE #83: Ability to bind Mouse Buttons as keybindings

### PR DESCRIPTION
**Pull request by commy2**
**When merged this pull request will:**
- remakes the key editing ui for CBA keys to support mouse buttons and multiple keybinds
- the ui is made to be as close to the vanilla ui as possible and behaves the same
- close #83

Also has all the features the vanilla keybinding ui has. Undo, Delete, Default, Next, Prev and they all do exactly the same.

### CODE REVIEW

**Comment by commy2** [comment]
```sqf
[
    "Test Category",
    "Test",
    ["key action name", "key action tooltip"],
        {systemChat str 1},
        {systemChat str 2},
    [15, [false, false, false]]
] call cba_fnc_addKeybind;
```

**Comment by commy2** [comment]
note:
- split tweaks for `CBA_fnc_addKeybind` from this to keep PR small...
- same for settings ui tweaks (copy pasted tweaks from this)?

File: addons/keybinding/fnc_addKeybind.sqf [hunk 1]
```diff
@@ -2,170 +2,147 @@
...
+private _keybind = [_key, [_shift, _control, _alt]];
 
-_entryIndex = (GVAR(defaultKeybinds) select 0) find _hashDown;
-if(_entryIndex == -1) then {
-    _entryIndex = (GVAR(defaultKeybinds) select 0) pushBack _hashDown;
-    (GVAR(defaultKeybinds) select 1) set[_entryIndex, []];
+// data for controls menu
+if (isNil QGVAR(actions)) then {
+    GVAR(actions) = [] call CBA_fnc_createNamespace;
+    //GVAR(keys) = [] call CBA_fnc_createNamespace;
```
Comment (Killswitch00) [inline addons/keybinding/fnc_addKeybind.sqf]: Not needed?

Comment (commy2) [inline addons/keybinding/fnc_addKeybind.sqf]: Yeah, it's all still experimental and wip.

**Comment by commy2** [comment]
Part of this went into #656, so this PR can eventually only change the gui and not `CBA_fnc_addKeybind`.

**Comment by commy2** [comment]
todo: rebase to keybinding

**Comment by commy2** [comment]
done

**Comment by PabstMirror** [comment]
ACE uses -1 for some unbound keys
```
["Pabst", "test2", "test2",{systemChat "test2";}, {}, [-1, [false, false, false]], false] call CBA_fnc_addKeybind;
```
throws
```
Error position: <set [_key, _hashKeys];
  Error Zero divisor
File \x\cba\addons\events\fnc_addKeyHandler.sqf [CBA_fnc_addKeyHandler], line 2014
```

File: addons/keybinding/fnc_addKeybind.sqf [hunk 2]
```diff
@@ -92,6 +92,8 @@ if (_defaultKeybind isEqualTypeParams [0, false, false, false]) then {
 _defaultKeybind params [["_defaultKey", 0, [0]], ["_defaultModifiers", [], [[]]]];
 _defaultModifiers params [["_defaultShift", false, [false]], ["_defaultControl", false, [false]], ["_defaultAlt", false, [false]]];
 
+_defaultKey = _defaultKey max 0;
```
Comment (PabstMirror) [inline addons/keybinding/fnc_addKeybind.sqf]: What about big numbers?
`min 254`?

Comment (commy2) [inline addons/keybinding/fnc_addKeybind.sqf]: Why would anyone use these?

File: addons/keybinding/XEH_preStart.sqf [hunk 3]
```diff
@@ -178,16 +178,18 @@ GVAR(keyNamesHash) = [_supportedKeys] call CBA_fnc_hashCreate;
...
+    [0xF2, 0x10002], // MMB
+    [0xF3, 0x10003], // Mouse #4
+    [0xF4, 0x10004], // Mouse #5
+    [0xF5, 0x10005], // Mouse #6
+    [0xF6, 0x10006], // Mouse #7
+    [0xF7, 0x10007], // Mouse #8
+    [0xF8, 0x100004], // Mouse wheel up
+    [0xF9, 0x100005], // Mouse wheel down
+    [84, DIK_SYSRQ],
+    [198, DIK_PAUSE]
```
Comment (commy2) [inline addons/keybinding/XEH_preStart.sqf]: Alt+Print and Ctrl+Pause leads to UNEXPECTED_KEY, even in vanilla, but this makes them display the right key.

**Pull request merged**